nomos-libp2p/src/behaviour/nat/mod.rs

```diff
@@ -1,37 +1,111 @@
-use std::task::{Context, Poll};
+use std::{
+    task::{Context, Poll},
+    time::Duration,
+};
 
+use either::Either;
+use futures::{
+    future::{BoxFuture, Fuse, OptionFuture},
+    FutureExt as _,
+};
 use libp2p::{
-    autonat::v2::client::{Behaviour, Config},
+    autonat,
     core::{transport::PortUse, Endpoint},
     swarm::{
-        ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour, THandler, THandlerInEvent,
-        THandlerOutEvent, ToSwarm,
+        behaviour::toggle::{Toggle, ToggleConnectionHandler},
+        ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour, NewListenAddr, THandler,
+        THandlerInEvent, THandlerOutEvent, ToSwarm,
     },
     Multiaddr, PeerId,
 };
 use rand::RngCore;
 
+mod address_mapper;
+mod state_machine;
+
+use state_machine::{Command, StateMachine};
+use tokio::sync::mpsc::UnboundedReceiver;
+
+use crate::{behaviour::nat::address_mapper::AddressMapperBehaviour, AutonatClientSettings};
+
+type Task = BoxFuture<'static, Multiaddr>;
+
 /// This behaviour is responsible for confirming that the addresses of the node
 /// are publicly reachable.
 pub struct NatBehaviour<R: RngCore + 'static> {
+    /// The static public listen address is passed through this variable to
+    /// the `poll()` method. Unused if the node is not configured with a static
+    /// public IP address.
+    static_listen_addr: Option<Multiaddr>,
```

Comment:
Can't all these properties be grouped inside an enum that is given to the `NatBehaviour`? Instead of using `Option`s and `Toggle`s, we can have a variant which contains these things.